envoy/formatter/substitution_formatter_base.h

```diff
@@ -62,8 +63,30 @@ template <class FormatterContext> class FormatterProviderBase {
                          const StreamInfo::StreamInfo& stream_info) const PURE;
 };
 
+template <> class FormatterProviderBase<void> {
+public:
+  virtual ~FormatterProviderBase() = default;
+
+  /**
+   * Format the value with the given stream info.
+   * @param stream_info supplies the stream info.
+   * @return absl::optional<std::string> optional string containing a single value extracted from
+   *         the given stream info.
```

Comment:
when does it return nullopt?

Reply:
It depends on the implementation of the provider. But it most cases, it will return a nullopt if no valid value is extracted. Take the %UPSTREM_CLUSTER% formatter as an example, if there is no target cluster exist, then the formatter will return nullopt. And then empty string or "-" will be used in the final output.

This is completely same with the exist formatWithContext except this method only take a single stream info ref as input parameter.